hilti/toolchain/src/compiler/optimizer.cc

```diff
@@ -1415,6 +1472,563 @@ struct MemberVisitor : OptimizerVisitor, visitor::PreOrder<bool, MemberVisitor>
+        }
+
+        return false;
+    }
+
+    result_t operator()(const declaration::Field& x, position_t p) {
+        const Function* fn = nullptr;
+        auto path = p.path;
+
+        if ( const auto& inline_ = x.inlineFunction() ) {
+            // Field already marked pure.
+            if ( const auto& attrs = inline_->attributes(); attrs && attrs->find("&pure") )
+                return false;
+
+            fn = &*inline_;
+            path.emplace_back(const_cast<Node&>(p.node.children()[3]));
+        }
+
+        else {
+            if ( const auto& type = p.parent(2).tryAs<declaration::Type>() ) {
+                if ( auto function_ =
+                         scope::lookupID<declaration::Function>(ID(util::fmt("%s::%s", type->canonicalID(), x.id())), p,
+                                                                "function") ) {
+                    const auto& function = function_->first->as<declaration::Function>();
+
+                    fn = &function.function();
+                    path.emplace_back(const_cast<Node&>(function.children()[0]));
+                }
+            }
+        }
+
+        if ( fn ) {
+            assert(! path.empty());
+            auto& node = path.rbegin()->node;
+            auto pn = position_t{node, path};
+
+            if ( auto new_function = pureFunction(*fn, pn) ) {
+                HILTI_DEBUG(logging::debug::Optimizer,
+                            util::fmt("marking member '%s' as pure (%s)", fn->id(), fn->meta().location()));
+
+                node = new_function.value();
+                return true;
+            }
+        }
+
+        // Methods not declared inline are handled via their function declaration.
+
+        return false;
+    }
+
+    // Helper function to which either computes a pure version of the given
+    // input function, or nothing if it cannot detect that the functions is
+    // pure.
+    std::optional<Function> pureFunction(const Function& x, position_t p) const {
+        if ( const auto& attrs = x.attributes() ) {
+            // The function is already marked pure.
+            if ( attrs->find("&pure") )
+                return {};
+
+            // We cannot analyze functions implemented in C++ so they cannot be pure.
+            if ( attrs->find("&cxxname") )
+                return {};
+        }
+
+        // We cannot analyze functions for which we have no body.
+        if ( ! x.body() )
+            return {};
+
+        // We determine whether the function can be pure by looking at all the
+        // IDs it references and members it calls. We assume this function is
+        // pure unless we can find a reason it should not be.
+        auto v = visitor::PreOrder<>();
+        for ( const auto& child : v.walk(p.node) ) {
+            if ( auto id = child.node.tryAs<expression::ResolvedID>() ) {
+                const auto& decl = id->declaration();
+
+                if ( auto local = decl.tryAs<declaration::LocalVariable>() ) {
```

Comment:
I'm wondering if we have more cases where an access to a local is not pure. I can't think of much right now, but one case: what about locals of reference type?